### PR TITLE
Convert Imports to Indirect Calls for Dynamic Linking

### PR DESCRIPTION
**Pull request by awtcode**
This complements the work done in https://github.com/WebAssembly/binaryen/pull/2523 to convert Imports in Dynamic Linking into Indirect Calls. The benefits of doing this are:

1. Faster performance
2. Skip legalization
3. Remove JS stubs (To be done in a later PR)

A sample benchmark has been uploaded here: https://github.com/awtcode/dylink_benchmark

These are the results of the benchmark:

| Chrome 78  |  |
| ------------- | ------------- |
| Original  | 227ms  |
| PR  | 17ms  |

| Firefox 71  |  |
| ------------- | ------------- |
| Original  | 43ms  |
| PR  | 6ms  |

This PR gives a 13x improvement in Chrome and a 7x improvement in FF.

### CODE REVIEW

**Comment by awtcode** [comment]
@sbc100 , this is my initial attempt to replace imports with indirect calls. Can you take a look and give me your comments? Thanks.

File: emscripten.py [hunk 1]
```diff
@@ -1800,8 +1800,10 @@ def create_fp_accessors(metadata):
     # the name of the original function is generally the normal function
     # name, unless it is legalized, in which case the export is the legalized
     # version, and the original provided by orig$X
+    originalAccessor = ''
     if shared.Settings.LEGALIZE_JS_FFI and not shared.JS.is_legal_sig(sig):
       name = 'orig$' + name
+      originalAccessor = ('if (!func) func = Module["_%s"];' % (name))
```
Comment (sbc100) [inline emscripten.py]: Remove the extra pair of (  ) around this statement.

File: emscripten.py [hunk 2]
```diff
@@ -1810,13 +1812,14 @@ def create_fp_accessors(metadata):
   var func = Module['asm']['%(original)s'];
   // If there is no wasm function, this may be a JS library function or
   // something from another module.
+  %(originalAccessor)s
```
Comment (sbc100) [inline emscripten.py]: I think this part is already handles above now isn't it?

I think that was what alon fixed in #9810?  Maybe this part of this change can be reverted now?

File: emscripten.py [hunk 3]
```diff
@@ -2176,8 +2179,24 @@ def emscript_wasm_backend(infile, outfile, memfile, compiler_engine,
...
+    glue, forwarded_data = compile_settings(compiler_engine, temp_files)
+    forwarded_json = json.loads(forwarded_data)
+    library_fns = forwarded_json['Functions']['libraryFunctions']
+    library_fns_list = []
+    for name in library_fns:
+      library_fns_list.append(name.encode('utf-8'))
+
+    libraryfile = infile + '.libfile'
+    with open(libraryfile, 'w') as f:
+      f.write(str(library_fns_list))
```
Comment (sbc100) [inline emscripten.py]: I better format would probably just one line per symbols:

```
for sym in library_fns_list:
   f.write(sym + '\n')
```

Comment (sbc100) [reply emscripten.py]: Can we call the file, and this variable somehow more meaningful too? Perhaps "js_imports"?  Or "js_symbols"?

File: emscripten.py (hunk 2 above)
Comment (awtcode) [reply emscripten.py]: I believe the side exports are added to the `Module` object and not `Module['asm']` thus the line `var func = Module['asm']['%(original)s'];` can only retrieve the original exports from the Main module and not the side. For this PR, we actually need the original functions from the side module, thus the need to retrieve the original side functions from the `Module` object instead  `originalAccessor = ('if (!func) func = Module["_%s"];' % (name))`

File: emscripten.py (hunk 3 above)
Comment (awtcode) [reply emscripten.py]: Thanks for the suggestion, will update the format :)

**Comment by awtcode** [comment]
@sbc100 and @kripken , I have updated the PR with some results from my test app.

**Comment by kripken** [comment]
I am uncomfortable with getting the list of JS library functions in this way. It means that there are two kinds of imports, one that is provided by wasm and one by JS, and the module being compiled needs to know which is which at compile time. This feels a little risky.

IIUC this PR makes imported wasm functions be called indirectly. If that is fast enough, then making the JS functions be indirect calls as well may be fast enough too. That is, JS library functions would be added to the table, then treated just like any other import. That layer could also add legalization for JS as needed, which would simplify our legalization story.

(However, that assumes these indirect calls are fast enough, and we may want to keep one direction's calls as direct, for speed? That would make this more complicated.)

File: emscripten.py (hunk 1 above)
Comment (awtcode) [reply emscripten.py]: Sure

File: emscripten.py [hunk 4]
```diff
@@ -2206,8 +2206,25 @@ def emscript_wasm_backend(infile, outfile, memfile, compiler_engine,
...
+    library_fns_list = []
+    for name in library_fns:
+      library_fns_list.append(name)
+
+    js_symbols = infile + '.js_symbols'
+    with open(js_symbols, 'w') as js_symbols_file:
+      for sym in library_fns_list:
+        js_symbols_file.write(sym + '\n')
+  else:
+    js_symbols = []
```
Comment (sbc100) [inline emscripten.py]: I stole this idea/code and made it part of #10350 in a function called `get_all_js_library_funcs` hopefully that is about to land and we can use that shared function here.

File: tests/test_core.py [hunk 5]
```diff
@@ -8559,6 +8559,41 @@ def test_safe_stack_dylink(self):
...
+      int sidey() {
+        printf("sidey says %d ", 17);
+        return 17;
+      }
+
+      int sidey2() {
+        printf("sidey2 says %d", 18);
+        return 18;
+      }
+    ''', expected='sidey says 17 sidey2 says 18', need_reverse=False)
```
Comment (sbc100) [inline tests/test_core.py]: Can we somehow verify that this worked by looking at import/exports of the resulting wasm file?

**Pull request closed without merging**